### PR TITLE
fix(appointment): correct last day boundary calculation

### LINKED ISSUE #2959: Monthly appointment fetch excludes the last day of the month (claude assist)
Monthly appointment fetch **excludes appointments on the last day of the month**. In `AppointmentManagerImpl.findMonthlyAppointments` (~lines 281-313), the end date is set to the last day of the month, then `HOUR/MINUTE/SECOND` are forced to 0 and `cal.add(SECOND, -1)` rolls it back to the **previous day 23:59:59**, so `findByDateRangeAndProvider` never includes the actual last calendar day. (The code comment on the adjacent line even warns about exactly this.)

**Fix direction:** set the end boundary to the last day at 23:59:59 (or use an exclusive next-day start).

**Suggested type:** Bug · **Priority:** 3-Medium

---
Found during the API testing pass in #2916 (verified against `develop`, 2026-06-19). Filed with Claude assistance.

### PR DESCRIPTION
**Pull request by ThanuH**
## Description

Fixes a bug in AppointmentManagerImpl.findMonthlyAppointments where appointments scheduled on the last day of the month were silently excluded from results.

The previous logic attempted to calculate an inclusive upper bound for the month, which conflicted with the OscarAppointmentDaoImpl.findByDateRangeAndProvider method's reliance on an exclusive upper bound (appointmentDate < endTime). By attempting to hit the final second of the month, the code failed to capture appointments on the final day.

**Changes made:**
-Changed endDate calculation to the first moment of the subsequent month (00:00:00.000).

-This utilizes the DAO's exclusive boundary contract to ensure every millisecond of the last day is covered.

-Zeroed out MILLISECOND on the startDate to ensure consistent range start-points.

-Removed the legacy cal.add(SECOND, -1) logic and updated comments to reflect the exclusive boundary pattern.

## Related Issues

#2959 

## How Was This Tested?

Manually verified the fix by calling `findMonthlyAppointments` for a month with appointments on the last day (e.g. January 31, March 31) and confirming they are now included in the returned list.

Tested edge cases:
- Month ending on the 28th (February, non-leap year)
- Month ending on the 29th (February, leap year)
- Month ending on the 30th (April, June, September, November)
- Month ending on the 31st (January, March, May, July, August, October, December)

Confirmed start date is `00:00:00.000` and end date is `23:59:59.999` on the correct boundary days via the existing `logger.info` output.

## Screenshots

N/A — no visual changes.

## Checklist
- [ x] My commits are signed off for the [DCO](https://developercertificate.org/) (`git commit -s`)
- [ x] My commits follow [Conventional Commits](https://www.conventionalcommits.org/) format, or I've written clear commit messages and will use the format next time
- [ x] I have not included any patient data (PHI) in this PR
- [x ] I have added tests for new functionality, or this change doesn't need new tests
- [ x] I have read the [contributing guide](CONTRIBUTING.md)

## Summary by Sourcery

Fix monthly appointment range calculation to correctly include appointments on the last calendar day of the month.

Bug Fixes:
- Ensure the monthly appointment query end date covers the full final day of the month instead of stopping one second before midnight of the previous day.

Enhancements:
- Normalize start and end date millisecond fields when computing the monthly appointment date range to avoid partial-second offsets in queries.

---
## Summary by cubic
Fix the monthly appointment query so appointments on the last calendar day are included. Use a half-open [start, end) range and add unit tests for month boundaries, including leap-year February.

- **Bug Fixes**
  - Start set to 00:00:00.000 with milliseconds cleared.
  - End set to next month’s 00:00:00.000 (exclusive); removed subtract-one-second logic and documented/logged the exclusive bound.

<sup>Written for commit 9d90d901b0b24edef4d87b695e3b0e153970596f. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/carlos-emr/carlos/pull/2970?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

## Summary by Sourcery

Correct monthly appointment date range calculation to align with the DAO’s exclusive upper-bound contract so final-day appointments are included.

Bug Fixes:
- Ensure monthly appointment queries include appointments scheduled on the last calendar day of the month by using an exclusive end date at the start of the next month.

Enhancements:
- Normalize start and end date times to zero milliseconds when computing the monthly appointment date range for consistent query boundaries.

### CODE REVIEW

File: src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java [hunk 1]
```diff
@@ -287,22 +287,18 @@ public List<Appointment> findMonthlyAppointments(LoggedInInfo loggedInInfo, Stri
...
-        cal.set(Calendar.HOUR_OF_DAY, 0);
-        cal.set(Calendar.MINUTE, 0);
-        cal.set(Calendar.SECOND, 0);
-
-        cal.add(Calendar.SECOND, -1);
+        cal.set(Calendar.HOUR_OF_DAY, 23);
+        cal.set(Calendar.MINUTE, 59);
+        cal.set(Calendar.SECOND, 59);
+        cal.set(Calendar.MILLISECOND, 999);
         Date endDate = cal.getTime();
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java]: Not a concern in this use case.

File: src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java [hunk 2]
```diff
@@ -287,22 +287,18 @@ public List<Appointment> findMonthlyAppointments(LoggedInInfo loggedInInfo, Stri
...
 
-
         logger.info("monthly - checking from " + startDate + " to " + endDate);
 
-
         List<Appointment> results = appointmentDao.findByDateRangeAndProvider(startDate, endDate, providerNo);
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java]: This will be covered in further permissions work outside this PR.

File: src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java [hunk 3]
```diff
@@ -287,22 +287,18 @@ public List<Appointment> findMonthlyAppointments(LoggedInInfo loggedInInfo, Stri
         cal.set(Calendar.HOUR_OF_DAY, 0);
         cal.set(Calendar.MINUTE, 0);
         cal.set(Calendar.SECOND, 0);
+        cal.set(Calendar.MILLISECOND, 0);
         Date startDate = cal.getTime();
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java:307]: Legacy issues; acceptable for now.

**Review by yingbull — DISMISSED** [review]

File: src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java [hunk 4]
```diff
@@ -287,22 +287,18 @@ public List<Appointment> findMonthlyAppointments(LoggedInInfo loggedInInfo, Stri
...
-
-        cal.set(Calendar.HOUR_OF_DAY, 0);
-        cal.set(Calendar.MINUTE, 0);
-        cal.set(Calendar.SECOND, 0);
-
-        cal.add(Calendar.SECOND, -1);
+        cal.set(Calendar.HOUR_OF_DAY, 23);
+        cal.set(Calendar.MINUTE, 59);
+        cal.set(Calendar.SECOND, 59);
+        cal.set(Calendar.MILLISECOND, 999);
```
[the remark this answers is not in the document]
Comment (yingbull) [reply src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java]: Acceptable in this use case.

File: src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java [hunk 5]
```diff
@@ -287,22 +287,18 @@ public List<Appointment> findMonthlyAppointments(LoggedInInfo loggedInInfo, Stri
...
-
-        cal.set(Calendar.HOUR_OF_DAY, 0);
-        cal.set(Calendar.MINUTE, 0);
-        cal.set(Calendar.SECOND, 0);
-
-        cal.add(Calendar.SECOND, -1);
+        cal.set(Calendar.HOUR_OF_DAY, 23);
+        cal.set(Calendar.MINUTE, 59);
+        cal.set(Calendar.SECOND, 59);
+        cal.set(Calendar.MILLISECOND, 0);
```
[the remark this answers is not in the document]
Comment (Ben-Heerema) [reply src/main/java/io/github/carlos_emr/carlos/managers/AppointmentManagerImpl.java]: Your reasoning is wrong but the fix looks cleaner than the current implimentation. Thanks

**Comment by Ben-Heerema** [comment]
@ThanuH The fix looks good but do you mind checking out copilots suggested fix to see if that's more reasonable? If not I can do it, just wanted to give you the satisfaction. Honestly, I am leaning towards copilot but I wouldn't mind hearing your thoughts :)

**Comment by ThanuH** [comment]
Hi @Ben-Heerema After reviewing Copilot's suggestion and considering the DAO method findByDateRangeAndProvider which uses an exclusive upper bound (appointmentDate < endTime), setting endDate to the first moment of the next month (00:00:00.000) is the more correct fix this ensures every millisecond of the last day is covered without any gaps.

i'll do the changes, thanks for the comment

**Comment by Ben-Heerema** [comment]
Hey @ThanuH, looks like these commits are missing DCO sign off: 51369d7 fix: update endDate calculation to use exclusive next-month boundary, caa155c Merge remote-tracking branch 'origin/fix/appointment-last-day-boundary' into fix/appointment-last-day-boundary. Do you mind amending those? If it's not easy to do I can silence them myself later. 

Also, try to keep the PR is draft unless you are ready for a review and all CI comments are resolved as unnecessary/dealt with.   Thanks for your work so far!

**Comment by ThanuH** [comment]
@Ben-Heerema Thanks for taking care of the DCO sign-offs on those commits, I really appreciate the help! I've marked the PR as ready for review. Let me know if there is anything else you need from me for this to get merged.

**Comment by Ben-Heerema** [comment]
@ThanuH It looks like the remaining CI is just adding more tests and cleaning up documentation. Good work and thanks for your patience but I can handle it from here if that's okay with you? Otherwise, if you are firm on doing this yourself check out the CI's comments. Thanks!

Note: I would usually be happy for you to finish it up but this specific PR is time critical as we would like to fix this before the next release, so if you're able to quickly fix it up I'd be happy to let you but if that's not reasonable for you it's totally cool and I will fix it up.

**Comment by ThanuH** [comment]
Hi @Ben-Heerema , thanks for the update and for the offer! Since this is time-critical for your release, I'm happy to let you take it from here so you can get it merged on time. I'm looking forward to the next one!

**Comment by Ben-Heerema** [comment]
@ThanuH Thanks for your help!

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

**Comment by Ben-Heerema** [comment]
@yingbull ready for review

**Comment by Ben-Heerema** [comment]
@coderabbitai review
@cubic-dev-ai review

**Review by yingbull — APPROVED** [review]

**Comment by yingbull** [comment]
@ThanuH thank you for your time  and involvement on this one! We'll have you shown as a coauthor to credit appropriately and appreciate it.
